### DISCUSSION TITLE
Deye SUN600G3-EU-230 with MW3_SSL_5408_1.0B - cannot get data from the inverter

### DISCUSSION THREAD

**Discussion by mario37359** [davidrapan/ha-solarman#132] [Q&A] [answered]
I tried to setup my Deye Inverter in  the HA with your ha-solarman extension, but cannot establish a connection. I tried several config files, the Inverter is reachable through the browser. Can you give me any hints on the issue or what to change in the config or parameters? Could not find anything helpless for my problem in the documentation or Q+A section. TY!
![e4_Screenshot 2024-09-02 151229](https://github.com/user-attachments/assets/946c34a7-b7f0-4fe3-94cf-3a1e81221aa4)
![e5_Screenshot 2024-09-02 151535](https://github.com/user-attachments/assets/47d5a20b-e0b5-48c1-bb6e-5119194d8fb4)
![e3_Screenshot 2024-09-02 112204](https://github.com/user-attachments/assets/b2538df0-9696-42b4-b1f4-a6f58d5cdf42)

**Comment by davidrapan** [comment]
You gonna have to give me debug logs from device addition flow:
1. Delete device
2. Enable debug by clicking on the "Debug Protokollierung aktivie..."
3. Try add device and wait for couple of minutes
4. Disable debug by clicking on the "Debug Protokollierung aktivie..."
5. Attach the file it just downloaded for you here in the thread.

😉

**Reply by mario37359** [reply]
Hej David, debug file, as wanted. Thank you!
[home-assistant_solarman_2024-09-02T17-38-18.502Z.log](https://github.com/user-attachments/files/16840197/home-assistant_solarman_2024-09-02T17-38-18.502Z.log)

**Reply by davidrapan** [reply]
Fuh, ok so you gonna have to ssh into your HA and run: `ha core rebuild`

**Reply by mario37359** [reply]
![e2_1_Screenshot 2024-09-03 073919](https://github.com/user-attachments/assets/e47551c0-d316-48b2-890c-0dc00b3d6c83)
I tried this and reinstalled, without success. Send the log to check again... TY
[home-assistant_solarman_2024-09-03T05-40-16.206Z.log](https://github.com/user-attachments/files/16843972/home-assistant_solarman_2024-09-03T05-40-16.206Z.log)

**Reply by davidrapan** [reply]
So initial problem has been fixed. Now you are maybe using wrong inverter profile..

What exact model of the inverter do you have?

**Reply by mario37359** [reply]
My current inverter is a Deye SUN600G3-EU-230

**Reply by davidrapan** [reply]
@mario37359 so you are just using incorrect profile, you can find the correct one here: https://github.com/davidrapan/ha-solarman/wiki/Supported-Devices

**Comment by PV-Joe** [comment]
I found something that could be related to the problem.

The problem seems to start with the Firmware: "MW3_SSL_5408_1.0B"

Unfortunately this is in German.

https://dy-support.org/community/deye-micro-inverter/port-48899-nach-ota-firmwareupdate-gesperrt/

Maybe it helps?

**Reply by davidrapan** [reply]
They are saying that just the 48899 for the discovery is closed but that it still normally works over 8899. 😉 

Edit: But nevertheless I really don't like what they are doing with those changes. Yeah it's just the discovery out for now but who knows what will be next... It can end up real bad for us...

**Reply by mario37359** [reply]
I'am in Germany, so the article is no problem for me 😉 I'll read about that.. TY

**Comment by davidrapan** [comment]
Guys @jmccrohan and @githubDante, I really don't like the new "MW3_SSL_5408_1.0B"...

```
RECD: a5 10 00 10 15 9a a3 be eb 42 fa 02 01 e6 4e 0c 00 7b 05 00 00 09 52 ca 66 05 00 aa 15
```
Doesn't look like modbus though...

Posts about the firmware and the issues with it (in German):
* https://dy-support.org/community/deye-micro-inverter/port-48899-nach-ota-firmwareupdate-gesperrt/
* https://www.photovoltaikforum.com/thread/233881-deye-firmware-mw3-ssl-5408-1-0b-kein-logging-mehr-m%C3%B6glich/
* https://www.solaranzeige.de/phpBB3/viewtopic.php?t=5023

**Reply by PV-Joe** [reply]
Carlo57 writes at the forum on solaranzeige.de (translated from German to English):

_I wrote an email to "servicede@deye.com.cn" yesterday, expressing my request for an open port 48899 or a firmware with this open port. Since today, the solar display is working again; I now have the firmware MW3_16U_5406_1.63 running._

Maybe this is a quick & dirty solution?

**Reply by davidrapan** [reply]
Well we don't have that firmware in hands so... 😆 

But! It's nice to see that they are willing to provide different versions on request. Sounds promising...

**Reply by PV-Joe** [reply]
> Well we don't have that firmware in hands so... 😆 

MW3_16U_5406_1.57 Firmware Download:

http://47.254.36.66/0_D0002_18/MW3_16U_5406_1.57.bin

**Reply by davidrapan** [reply]
😮

**Reply by davidrapan** [reply]
Uhm is the source safe?

**Reply by PV-Joe** [reply]
> Uhm is the source safe?

This is a URL from the Deye Firmware Update Server.

Unfortunately this version has significant security issues. So it is better to use a later version:

# Deye
http://47.254.36.66/0_D0002_18/MW3_15U_5406_1.471.bin
http://47.254.36.66/0_D0002_18/MW3_15U_5406_1.49R.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_1.53.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_1.56.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_1.57.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_2.04.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_2.06.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_2.30.bin
http://47.254.36.66/0_D0002_18/MW3_16U_5406_2.31.bin

# Solis Ginglong
http://47.254.36.66/0_D0002_18/MW3_15_0501_1.21.bin
http://47.254.36.66/0_D0002_18/MW3_15_0501_1.23.bin

# Deye Hybrid Inverters
http://47.254.36.66/0_5407_18/MCU1-SG-Ver0000-C362-CK.bin

**Reply by githubDante** [reply]
> ```
> RECD: a5 10 00 10 15 9a a3 be eb 42 fa 02 01 e6 4e 0c 00 7b 05 00 00 09 52 ca 66 05 00 aa 15
> ```
> 
> Doesn't look like modbus though...

It's an exception. The request is for register 244  in deye_hybrid.yaml  - [here](https://github.com/davidrapan/ha-solarman/blob/9015fba0b129db04c4b7a4fd996fa9ccb36e48fb/custom_components/solarman/inverter_definitions/deye_hybrid.yaml#L542). The address(es) appear(s) to be wrong.

**Reply by davidrapan** [reply]
@githubDante, thanks! So he just used wrong profile.

Edit: But still, I don't like that closed 48899..

**Reply by davidrapan** [reply]
@PV-Joe, ahahah, okay. Good to know!

**Reply by davidrapan** [reply]
@PV-Joe, could you also start new discussion with those firmware links and all the info you have about it? Thanks in advance. 😉

**Reply by PV-Joe** [reply]
> @PV-Joe, could you also start new discussion with those firmware links and all the info you have about it? Thanks in advance. 😉

Go here:
https://github.com/davidrapan/ha-solarman/discussions/135

**Reply by davidrapan** [reply]
> It's an exception.

Maybe a dumb question, but does that mean that we do know what these exceptions look like then? So maybe we can do log them better than just saying the frame is not a modbus frame?

**Reply by githubDante** [reply]
> > It's an exception.
> 
> Maybe a dumb question, but does that mean that we do know what these exceptions look like then? So maybe we can do log them better than just saying the frame is not a modbus frame?

The frame is invalid, that's for sure,  as it does not follow the MODBUS protocol specification. It's a plain exception code without slave address and CRC and as such the frame is shorter. Even the exception code which they use is incorrect

>0x05 - ACKNOWLEDGE - The server (or slave) has accepted the request and is processing it, but a long duration of time will be required to do so. This response is returned to prevent a timeout error from occurring in the client (or master). 

It  should be 0x02 for illegal data address.

Probably we can enrich the exception message a little bit in PySolarman by adding extra info in it as the exception can be checked easily with umodbus

```python
from umodbus.exceptions import error_code_to_exception_map
error_code_to_exception_map.get(v5_frame[25]).__name__
'AcknowledgeError'
```

**Reply by davidrapan** [reply]
> Even the exception code which they use is incorrect

I'm not surprised at all... lol.

So something like this? (pysolarmanv5.py at 258):
```
if len(modbus_frame) < 5:
    # Check to some min length or maybe there is some better way to check if even contains any data at all?
    # If it's needed at all though...
    if len(modbus_frame) > 0:
        modbus_exception = error_code_to_exception_map.get(modbus_frame[0]).__name__
        raise V5FrameError(f"V5 frame contains Modbus exception: {modbus_exception}")
    raise V5FrameError("V5 frame does not contain a valid Modbus RTU frame")
```

**Comment by davidrapan** [comment, marked as answer]
@mario37359, use this profile: deye_2mppt.yaml during the configuration. 😉

**Reply by mario37359** [reply]
Oh..that really works ...  👌 Thank you so much guys! 
![e2_Screenshot 2024-09-03 133132](https://github.com/user-attachments/assets/208d73dc-2a92-4723-b605-0844db7eb1ac)

**Comment by PV-Joe** [comment]
Congratulations!
Now you can start creating spicy dashboards 😄

If you like you can share them here and give some inspiration for others:

https://github.com/davidrapan/ha-solarman/discussions/new?category=show-and-tell
